partiql-lang/src/main/kotlin/org/partiql/lang/eval/visitors/FromSourceAliasVisitorTransform.kt

```diff
@@ -15,27 +15,44 @@ import org.partiql.pig.runtime.SymbolPrimitive
  * If provided with a query that has all of the from source aliases already specified, an exact clone is returned.
  */
 class FromSourceAliasVisitorTransform : VisitorTransformBase() {
+    // When this visitor reaches a top-level FromSource, it transforms it with a fresh instance
+    // of the helper visitor below.
+    // Currently, there are two places where a top-level FromSource can occur: in a SELECT and in a DML statement.
 
+    override fun transformExprSelect_from(node: PartiqlAst.Expr.Select): PartiqlAst.FromSource {
+        val newFrom = super.transformExprSelect_from(node) // this applies the transformation to any eligible subqueries
+        return InnerFromSourceAliasVisitorTransform().transformFromSource(newFrom)
+    }
+
+    override fun transformStatementDml_from(node: PartiqlAst.Statement.Dml): PartiqlAst.FromSource? {
```

Comment:
FromSource is recursive in the AST, as needed for complex joins.  The outer visitor here must be careful to catch only *top-level* FromSources, leaving it to the inner visitor to traverse their recursive structure.  

This is subtle enough, so that I forgot about it by now!  I'll tweak the comments to make this more clear.

Reply:
Makes sense!